scripts/us_census/pep/monthly_population_estimate/preprocess.py

```diff
@@ -378,30 +410,289 @@ def process(self):
         calls defined methods to clean, generate final
         cleaned CSV file, MCF file and TMCF file.
         """
-        for file in self._input_files:
+        #input_path = FLAGS.input_path
+        ip_files = os.listdir(self.input_path)
+        self.input_files = [
+            self.input_path + os.sep + file for file in ip_files
+        ]
+        if len(self.input_files) == 0:
+            logging.info("No files to process")
+            return
+        processed_count = 0
+        total_files_to_process = len(self.input_files)
+        logging.info(f"No of files to be processed {len(self.input_files)}")
+        for file in self.input_files:
             df = self._load_data(file)
-            self._transform_data(df)
-        self._generate_mcf(self._df.columns)
-        self._generate_tmcf(self._df.columns)
+            result = self._transform_data(df)
+            if result:
+                processed_count += 1
+            else:
+                logging.fatal(f'Failed to process {file}')
+        logging.info(f"No of files processed {processed_count}")
+        if total_files_to_process > 0 & (processed_count
+                                         == total_files_to_process):
+            self._generate_mcf(self._df.columns)
+            self._generate_tmcf(self._df.columns)
+        else:
+            logging.fatal(
+                "Aborting output files as no of files to process not matching processed files"
+            )
 
 
-def main(_):
-    input_path = FLAGS.input_path
+def add_future_year_urls():
+    """
+    This method scans the download URLs for future years.
+    """
+    global _FILES_TO_DOWNLOAD
+    with open(os.path.join(_MODULE_DIR, 'input_url.json'), 'r') as inpit_file:
+        _FILES_TO_DOWNLOAD = json.load(inpit_file)
+    urls_to_scan = [
+        "https://www2.census.gov/programs-surveys/popest/tables/2020-{YEAR}/national/totals/NA-EST{YEAR}-POP.xlsx"
+    ]
+    # This method will generate URLs for the years 2021 to 2029
+    # need to the latest avaibale year
+    for future_year in range(2030, 2021, -1):

```

Comment:
Pls note that range(2030, 2021, -1) will stop at 2022.
I believe the most recent file `/2020-2023/` alone should be sufficient if it includes all years.
the drop_duplicates() may be dropping the data from other files.

Reply:
To identify the most recent available URL between 2023 and 2029, we're iterating through the years in reverse order, starting from 2030. Once the latest accessible URL is found, the process terminates. This approach has been thoroughly tested and is expected to function correctly.
drop_duplicate() is for year 2020 as the data found across two URLs .overlapping data found in the source so this drop duplicate is required